### PR TITLE
Dev clean dstl 20180928

### PR DESCRIPTION
**Pull request by mstefferson**
A former fellow, Ben Hammel, suggested pre-training on another dataset. Starting from his code, I have built out a pipeline to generate a nice dataset for pretraining.  This branch is adding this pipeline. Some of the doc string and formats are messed up (from the code I didn't write), but I don't have time to fix it now.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,7 +15,10 @@ notes.txt
 *pkl
 *log
 *jpg
+*tif
+*xml
 notebooks/Scratch.ipynb
+Untitled.ipynb
```
Comment (agude) [inline .gitignore:21]: May want to make this more general like `Untitled*ipynb`.

File: README.md [hunk 2]
```diff
@@ -9,62 +9,64 @@ Framework for machine learning projects at Insight Data Science.
...
-git checkout -b $branch_name
-git push origin $branch_name
+git clone https://github.com/mstefferson/AutomatedTreeCensus
 ```
 
-## Requisites
-- List all packages and software needed to build the environment
-- This could include cloud command line tools (i.e. gsutil), package managers (i.e. conda), etc.
+## Building info
+In order to interact with satellite data (both GIS and tif images), we need to use GDAL (Geospatial Data Abstraction Library). I found the installation of this to be a bit of a pain, so I build a docker image. With the exception of training on a AWS GPU node, all code should be ran through the docker image. More details on training vs processing below.
```
Comment (agude) [inline README.md:19]: "built a docker image"

File: README.md [hunk 3]
```diff
@@ -9,62 +9,64 @@ Framework for machine learning projects at Insight Data Science.
...
+
+This docker image is built from thinkwhere/gdal-python. It contains GDAL and many geospatial python libraries. It also supports jupyter notebooks.
```
Comment (agude) [inline README.md:21]: Enclose thinkwhere/gdal-python in backticks: `thinkwhere/gdal-python`

File: README.md [hunk 4]
```diff
@@ -9,62 +9,64 @@ Framework for machine learning projects at Insight Data Science.
...
-## Requisites
-- List all packages and software needed to build the environment
-- This could include cloud command line tools (i.e. gsutil), package managers (i.e. conda), etc.
+## Building info
+In order to interact with satellite data (both GIS and tif images), we need to use GDAL (Geospatial Data Abstraction Library). I found the installation of this to be a bit of a pain, so I build a docker image. With the exception of training on a AWS GPU node, all code should be ran through the docker image. More details on training vs processing below.
+
+This docker image is built from thinkwhere/gdal-python. It contains GDAL and many geospatial python libraries. It also supports jupyter notebooks. 
+
+## Docker
+With the exception of GPU training, building the docker image, and starting up a Docker container, all code should be ran within a Docker container. The docker images mounts the WD into the container.
```
Comment (agude) [inline README.md:24]: WD?

File: README.md [hunk 5]
```diff
@@ -9,62 +9,64 @@ Framework for machine learning projects at Insight Data Science.
...
 ```
-# Example
 
-# Step 1
-# Step 2
-```
+Note, the docker container is removed when exiting
+
+# Data
+The raw data I was given was a single tif file with four channel's (R, G, B, IR) of all of Athens. In order to train, I did pretrained on Kaggle's DSTL
```
Comment (agude) [inline README.md:52]: What is DSTL?

File: src/preprocess/clean_dstl/build_dstl.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
+import clean_dstl
+import label_dstl
+import argparse
```
Comment (agude) [inline src/preprocess/clean_dstl/build_dstl.py:3]: Great!

File: src/preprocess/clean_dstl/build_dstl.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        config (dist): loaded dstl config json. Contains image and path info
+    Returns:
+        N/A
+    Update:
+        N/A
+    Writes to file:
+        Writes /path/2/processed/data/(train, val)/(images, labels)
+    '''
+    # make dirs
+    if not os.path.exists(os.getcwd() + config["dstl"]["proc_data_rel"]):
```
Comment (agude) [inline src/preprocess/clean_dstl/build_dstl.py:25]: Use `os.path.join` instead of concat.

File: src/preprocess/clean_dstl/build_dstl.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        )
+    print('Processed all data')
+    # get all the bound box labels
+    df, files2delete = label_dstl.get_all_bounding(
+        config['dstl']['proc_data_rel'], config['dstl']['imag_w'],
+        config['dstl']['imag_h'])
+    print('Got all bounding boxes')
+    # get all the bound box labels
+    label_dstl.build_labels(df, files2delete, config['dstl']['imag_w'],
+                            config['dstl']['imag_h'],
```
Comment (agude) [inline src/preprocess/clean_dstl/build_dstl.py:55]: You are using multiple styles for indenting.

File: src/preprocess/clean_dstl/data_utils.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+class RawObjImage(pipe_utils.SatelliteImage):
+    """
+    """
+
+    def __init__(self, image_path):
+        self._image_id = pipe_utils.get_file_name_from_path(image_path)
+
+        # import the data for the image and the json
+        print("loading image ", self._image_id)
+        _data = read_raw_image(image_path)
```
Comment (agude) [inline src/preprocess/clean_dstl/data_utils.py:27]: Python uses 2 _ for "private" variables.

File: src/preprocess/clean_dstl/data_utils.py [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+    def append_feature(self, label, coor):
+        """Connect a feature to the imported image
+
+        Args
+        ----
+        label (str) : label of the feature
+        """
+        self._features.setdefault(label, []).append(coor)
+
+    def has_labels(self):
```
Comment (agude) [inline src/preprocess/clean_dstl/data_utils.py:46]: This name makes me think it should return `True` if it has labels.

File: src/preprocess/clean_dstl/data_utils.py [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+        return self._features
+
+    def append_feature(self, label, coor):
+        """Connect a feature to the imported image
+
+        Args
+        ----
+        label (str) : label of the feature
+        """
+        self._features.setdefault(label, []).append(coor)
```
Comment (agude) [inline src/preprocess/clean_dstl/data_utils.py:44]: This looks really hacky. Why is feature ever non-existant?

File: src/preprocess/clean_dstl/data_utils.py [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+
+    Returns
+    -------
+    a dic of the images of form { 'label': [ [img], ...], ...}
+    """
+    data = {}
+
+    for img_path in glob.glob(path + "/**/*"):
+
+        *_, label, im_name = img_path.split(os.path.sep)
```
Comment (agude) [inline src/preprocess/clean_dstl/data_utils.py:139]: Oh, tricky!

File: src/preprocess/clean_dstl/data_utils.py [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+                img = read_raw_image(img_path)
+            except ImportError:
+                # Assuming a file with an extension other than .jpg, .tif,
+                # or .png was in the directory, just skip
+                #
+                # WARNING: this is going on faith that read_raw_image wont
+                # raise an error for other reasons... that's probably not a 
+                # good assumption
+                pass
+            else:
```
Comment (agude) [inline src/preprocess/clean_dstl/data_utils.py:152]: Nice `try`/`except`/`else`!

File: src/preprocess/clean_dstl/data_utils.py [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+                img = pipe_utils.image_save_preprocessor(img)
+                data.setdefault(label, []).append(img)
```
Comment (agude) [inline src/preprocess/clean_dstl/data_utils.py:154]: You use this pattern a lot, but I haven't seen it before. That suggests to me there is a better way to handle it.

**Review by agude — COMMENTED** [review]
Ok, a bunch of comments, see inline! Sorry, I wasn't able to get through all of it in my self-allotted 30 minutes. ;-)
